tree_assume/src/type_analysis.egg

```diff
@@ -0,0 +1,165 @@
+(ruleset type-analysis)
+
+(relation HasType (Expr Type))
+
+;; Keep track of type expectations for error messages
+(relation ExpectType (Expr Type String))
+
+(rule (
+        (ExpectType e expected msg)
+        (HasType e actual)
+        (!= expected actual) ;; OKAY to compare types for equality because we never union types.
+      )
+      ((panic "type mismatch, better error message later :(")) ; once egglog allows non-literal panics
+      :ruleset type-analysis)
+
+; Primitives
+(rule ((= lhs (Const (Int i))))
+      ((HasType lhs (Base (IntT))))
+      :ruleset type-analysis)
+
+(rule ((= lhs (Const (Bool b))))
+      ((HasType lhs (Base (BoolT))))
+      :ruleset type-analysis)
+
+(rule ((= lhs (Empty)))
+      ((HasType lhs (TupleT (TNil))))
+      :ruleset type-analysis)
+
+; Unary Ops
+(rule (
+        (= lhs (Uop (Not) e))
+        (HasType e (Base (BoolT)))
+      )
+      ((HasType lhs (Base (BoolT))))
+      :ruleset type-analysis)
+(rule ((= lhs (Uop (Not) e)))
+      ((ExpectType e (Base (BoolT)) "(Not) expected bool, received {?}"))
+      :ruleset type-analysis)
+
+
+(rule ((= lhs (Uop (Print) e)))
```

Comment:
We should ensure that the child has some type, even though we don't care what type it is.
Would be great if we checked that it wasn't a tuple type but that's hard. add a TODO for non-tuple checking